### ISSUE TITLE
torch.compile ignores forward hooks after compilation is triggered

### ISSUE THREAD

**Issue by baldassarreFe** [pytorch/pytorch#117758]
### 🐛 Describe the bug

Issue: when wrapping a model with `torch.compile()`, hooks that are registered after compilation is triggered will not run.

Expected outcome: hooks should run in both cases, or at least there should be a warning when attempting to register a hook on a compiled module.

Note: it's not about when `torch.compile()` is called but when the first compilation is triggered.

Code
```python
from functools import partial

import torch
import torch.nn as nn

device = torch.device("cuda", 0)
x = torch.randn(1024, 64, device=device)

for trigger_compilation in [False, True]:
    print("trigger_compilation:", trigger_compilation)

    # Setup
    model = nn.Sequential(
        nn.Linear(64, 512),
        nn.ReLU(),
        nn.Linear(512, 512),
        nn.ReLU(),
        nn.Linear(512, 64),
        nn.ReLU(),
    )
    model.train()
    model.to(device)
    model = torch.compile(model)
    optimizer = torch.optim.SGD(model.parameters(), lr=0.5)

    # Trigger compilation
    if trigger_compilation:
        model(x)

    # Register hook
    @torch._dynamo.disable
    def hook(m, i, o, *, idx, dest):
        dest[idx] = o.sum()
    linears = [m for m in model.modules() if isinstance(m, nn.Linear)]
    handles = []
    dest = {}
    for idx, linear in enumerate(linears):
        h = linear.register_forward_hook(partial(hook, idx=idx, dest=dest))
        handles.append(h)

    # Run
    for _ in range(4):
        dest.clear()
        y = model(x)
        loss = torch.mean((x - y) ** 2)
        loss.backward()
        optimizer.step()
        optimizer.zero_grad()
        print(loss.item(), {k: v.item() for k, v in dest.items()})

    # Cleanup
    for h in handles:
        h.remove()
    
    print()
```

Output
```
trigger_compilation: False
1.0034619569778442 {0: -926.8182373046875, 1: 263.5154113769531, 2: -673.0859375}
1.0002923011779785 {0: -935.1256103515625, 1: -248.4486846923828, 2: -1108.0654296875}
0.9978930950164795 {0: -939.6043701171875, 1: -544.6873779296875, 2: -1406.814453125}
0.9959683418273926 {0: -941.3616943359375, 1: -700.7630004882812, 2: -1586.9423828125}

trigger_compilation: True
1.0027713775634766 {}
0.9998803734779358 {}
0.9976662397384644 {}
0.995843231678009 {}
```

### Versions

```
PyTorch version: 2.3.0.dev20240115
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.27.7
Libc version: glibc-2.31

Python version: 3.10.13 | packaged by conda-forge | (main, Dec 23 2023, 15:36:39) [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-1048-aws-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 12.3.107
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A100-SXM4-80GB
Nvidia driver version: 535.104.12
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Byte Order:                         Little Endian
Address sizes:                      46 bits physical, 48 bits virtual
CPU(s):                             96
On-line CPU(s) list:                0-95
Thread(s) per core:                 2
Core(s) per socket:                 24
Socket(s):                          2
NUMA node(s):                       2
Vendor ID:                          GenuineIntel
CPU family:                         6
Model:                              85
Model name:                         Intel(R) Xeon(R) Platinum 8275CL CPU @ 3.00GHz
Stepping:                           7
CPU MHz:                            2999.998
BogoMIPS:                           5999.99
Hypervisor vendor:                  KVM
Virtualization type:                full
L1d cache:                          1.5 MiB
L1i cache:                          1.5 MiB
L2 cache:                           48 MiB
L3 cache:                           71.5 MiB
NUMA node0 CPU(s):                  0-23,48-71
NUMA node1 CPU(s):                  24-47,72-95
Vulnerability Gather data sampling: Unknown: Dependent on hypervisor status
Vulnerability Itlb multihit:        KVM: Mitigation: VMX unsupported
Vulnerability L1tf:                 Mitigation; PTE Inversion
Vulnerability Mds:                  Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Retbleed:             Vulnerable
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Vulnerable
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves ida arat pku ospke

Versions of relevant libraries:
[pip3] lovely-numpy==0.2.10
[pip3] numpy==1.24.4
[pip3] torch==2.3.0.dev20240115
[pip3] torchmetrics==0.10.3
[pip3] torchvision==0.18.0.dev20240115
[pip3] triton==2.2.0
[conda] blas                      1.0                         mkl    conda-forge
[conda] libblas                   3.9.0            16_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            16_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            16_linux64_mkl    conda-forge
[conda] lovely-numpy              0.2.10             pyhd8ed1ab_0    conda-forge
[conda] mkl                       2022.2.1         h84fe81f_16997    conda-forge
[conda] numpy                     1.24.4          py310ha4c1d20_0    conda-forge
[conda] pytorch                   2.3.0.dev20240115 py3.10_cuda12.1_cudnn8.9.2_0    pytorch-nightly
[conda] pytorch-cuda              12.1                 ha16c6d3_5    pytorch-nightly
[conda] pytorch-mutex             1.0                        cuda    pytorch-nightly
[conda] torchmetrics              0.10.3             pyhd8ed1ab_0    conda-forge
[conda] torchtriton               2.2.0+e28a256d71           py310    pytorch-nightly
[conda] torchvision               0.18.0.dev20240115     py310_cu121    pytorch-nightly
```

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov

**Comment by vulkomilev** [comment]
I will take a look

**Comment by vulkomilev** [comment]
@baldassarreFe  I think it is from the version it works on 2.1.2

**Comment by xmfan** [comment]
I'm able to repro the issue. Tagging dynamo folks, do we have guards for the number of registered forward hooks?

**Comment by xmfan** [comment]
Tentatively assigning to @voznesenskym for dynamo + hooks

**Comment by anijain2305** [comment]
This is because TorchDynamo does not guard on nn module attributes by default. 

We can do this - `torch._dynamo.config.guard_nn_modules = True` to guard on nn module attributes. This will recompile the module on the second call because hooks are added between the first and second call. 

Note that this can introduce some (hopefully small) performance penalty because Dynamo will add lot of nn module guards.

**Comment by anijain2305** [comment]
Removing `high-priority` label as this is duplicate of already high pri issue - https://github.com/pytorch/pytorch/issues/111785

**Comment by anijain2305** [comment]
This is fixed. Closing.

**Comment by yuhuishi-convect** [comment]
hello @anijain2305 , is this fixed internally, or users need to set `torch._dynamo.config.guard_nn_modules = True` as a walkaround?
